clean_cap_dump.c

```diff
@@ -0,0 +1,245 @@
+		{ .netip = 0x000200c0, .netmask = 0x00ffffff  }, /* 192.0.2.0/24 */
+		{ .netip = 0x006433c0, .netmask = 0x00ffffff  }, /* 192.88.99.0/24 */
+		{ .netip = 0x0000a8c0, .netmask = 0x0000ffff  }, /* 192.168.0.0/16 */
+		{ .netip = 0x000012c0, .netmask = 0x0000fffe  }, /* 192.18.0.0/15 */
+		{ .netip = 0x006433c0, .netmask = 0x00ffffff  }, /* 198.51.100.0/24 */
+		{ .netip = 0x007100cb, .netmask = 0x00ffffff  }, /* 203.0.113.0/24 */
+		{ .netip = 0x000000e0, .netmask = 0x000000f0  }, /* 224.0.0.0/4 */
+		{ .netip = 0x000000f0, .netmask = 0x000000f0  }, /* 240.0.0.0/4 */
+		{ .netip = 0xffffffff, .netmask = 0xffffffff  }  /* 255.255.255.255 */
+	};
+	int sb_sz = sizeof(specialblock)/sizeof(specialblock[0]);
+	int reserved = 0;
+	int i = 0;
+	uint32_t addr = nd_ipv4_to_network_uint(a);
+
+	for (i = 0; i < sb_sz; i++) {
+		if ((addr & specialblock[i].netmask) == 
+		    (specialblock[i].netip & specialblock[i].netmask)) {
+			reserved = 1;
+			break;
+		}
+	}
+
+	return reserved;
+}
+
+/* Takes as input the IP header struct pointer and length of packet, and
+ * returns -1 if the IP header is parseable; else returns 0. */
+static int
+validate_iph_len(u_char *iph, unsigned int len) {
+	struct ip *ip = (struct ip *)iph;
+	if (IP_V(ip) != 4) {
+		printf("DEBUG: Not IP4 packet -- version %u\n", IP_V(ip));
+		return -1;
+	}
+
+	if (len < sizeof(struct ip)) {
+		printf("DEBUG: truncated IP or bad datagram length\n");
+		return -1;
+	}
+
+	if ((IP_HL(ip)*4) < sizeof(struct ip)) {
+		printf("DEBUG: bad header length\n");
+		return -1;
+	}
+	return 0;
+}
+
+/* Compare the addresses to reserved address ranges and mask if it's not within
+ * range. Returns 0 if successful, -1 on malformed datagram, 1 if not reserveds.
+ */
+int
+mask_ip(u_char *iph, unsigned int len, const char * maskIP) {
+	struct ip *ip = (struct ip *)iph;
+	uint32_t m = 0;
+	inet_pton(AF_INET, maskIP, &m);
+
+	if (validate_iph_len(iph, len) < 0)
+		return -1;
+
+	if (is_reserved(ip->ip_src) == 0) {
+		nd_ipv4 mask = network_uint_to_nd_ipv4(m);
+		memcpy(&(ip->ip_src), &mask, sizeof(mask));
+	}
+	if (is_reserved(ip->ip_dst) == 0) {
+		nd_ipv4 mask = network_uint_to_nd_ipv4(m);
+		memcpy(&(ip->ip_dst), &mask, sizeof(mask));
+	}
+
+	return 0;
+}
+
+void
+pcap_mod_and_dump(u_char *user, const struct pcap_pkthdr *h, const u_char *sp,
+	      int dlt, int no_payload_flag, int mask_ip_flag, const char *maskIP) {
+	register FILE *f;
+	struct clean_cap_sf_pkthdr sf_hdr;
+	u_char *ip, *p_end;
+	u_char *modp;
+	unsigned int p_len, modp_len;
+	p_len = modp_len = h->caplen;
+
+	if ((modp = (u_char *)malloc(MAXPACKET)) == NULL) {
+		fprintf(stderr,
+			"Unable to malloc %d bytes for packet modification", MAXPACKET);
+		exit(1);
+	} else
+		memset(modp, '\0', MAXPACKET);
+
+	memcpy(modp, sp, h->caplen);
+
+	f = (FILE *)user;
+	sf_hdr.ts.tv_sec    = h->ts.tv_sec;
+	sf_hdr.ts.tv_usec   = h->ts.tv_usec;
+	sf_hdr.caplen       = h->caplen;
+	sf_hdr.len          = h->len;
+
+	switch(dlt) {
```

Comment:
will there be more things processed by this switch? yes -> ok 
no -> probably should become an if